oxidation/libparsec/crates/types/src/regex.rs

```diff
@@ -0,0 +1,67 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+pub struct Regex(pub regex::Regex);
+
+/// The fnmatch_regex crate does not escape the character '$'. It is a problem
+/// for us. Since we need to match strings like `$RECYCLE.BIN` we need to escape
+/// it anyway.
+fn escape_globing_pattern(string: &str) -> String {
+    str::replace(string, "$", "\\$")
+}
+
+impl Regex {
+    pub fn from_pattern(pattern: &str) -> Result<Self, fnmatch_regex::error::Error> {
+        let escaped_str = escape_globing_pattern(pattern);
+        Ok(Regex(fnmatch_regex::glob_to_regex(&escaped_str)?))
+    }
+
+    pub fn from_regex_str(regex_str: &str) -> Result<Self, regex::Error> {
+        Ok(Regex(regex::Regex::new(regex_str)?))
+    }
+}
+
+impl PartialEq for Regex {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_string() == other.0.to_string()
+    }
+}
+
+mod tests {
+    #[test]
+    fn test_one_wildcard_regex() {
+        let r = crate::regex::Regex::from_pattern("*.rs").expect("Should be valid");
+
+        assert!(r.0.is_match("foo.rs"));
+        assert!(!r.0.is_match("bar.py"));
+    }
+
+    #[test]
+    fn test_two_wildcard_regex() {
+        let r = crate::regex::Regex::from_pattern("*bar*").expect("Should be valid");
+
+        assert!(r.0.is_match("foobar.rs"));
+        assert!(r.0.is_match("bar"));
+        assert!(!r.0.is_match("foo.py"));
+    }
+
+    #[test]
+    fn test_escape_dollar_in_pattern() {
+        let r = crate::regex::Regex::from_pattern("$myf$le.*").expect("Should be valid");
+        assert!(r.0.is_match("$myf$le.txt"));
+        assert!(!r.0.is_match("bar"));
+    }
+
+    #[test]
+    fn test_regex_str_creation() {
+        let r = crate::regex::Regex::from_regex_str(r"fooo?$").expect("Should be valid");
+        assert!(r.0.is_match("foo"));
+        assert!(r.0.is_match("fooo"));
+        assert!(!r.0.is_match("fooooooo"));
+    }
+
+    #[test]
+    fn test_bad_regex_str_creation() {
+        let r = crate::regex::Regex::from_regex_str(r"fooo][?");
+        assert!(r.is_err());
+    }
```

Comment:
Those tests are alike, so use `rstest` and convert those tests into `rstest::case`